clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -641,7 +643,7 @@ private void completeBatch(ProducerBatch batch, ProduceResponse.PartitionRespons
                 // thus it is not safe to reassign the sequence.
                 failBatch(batch, response, batch.attempts() < this.retries);
             }
-            if (error.exception() instanceof InvalidMetadataException) {
+            if (error.exception() instanceof InvalidMetadataException || error.exception() instanceof TimeoutException) {
```

Comment:
I wonder if this is necessary. It seems that the NetworkClient calls `metadataUpdater.handleServerDisconnect` when the node is disconnected and a metadata refresh is requested there. Could this be?

Reply:
I believe it is necessary and added a comment to the code to explain why. LMK if that makes sense. Thanks!

Reply:
I agree that we should refresh on `TimeoutException`. I wanted to point out that we are already doing it there: https://github.com/apache/kafka/blob/d6e35f3c41880c102faf774926cfb104a012de3f/clients/src/main/java/org/apache/kafka/clients/NetworkClient.java#L786. Hence I wonder if we need to add it here as well.

Reply:
@dajac For disconnected/timed out connections, this is the call stack:

```java
Sender.poll()
    NetworkClient.poll()
        DefaultMetadataUpdater.handleServerDisconnect()
            Metadata.requestUpdate()
````

However, `SenderTest` uses `MockClient` which doesn't have that same logic flow. So if I remove that clause of the `if` statement, the test, as written, fails.

Reply:
@kirktrue Do we need that new test if we remove the change in the Sender? We could perhaps remove it and keep the PR focused on the initial change. This would allow us to merge that first part.

Reply:
@dajac I removed the `error.exception() instanceof TimeoutException` check in `Sender` as you suggested. The unit test for ensuring that the metadata is updated on request timeouts has been added to the pre-existing request timeout tests in `NetworkClientTest`.